solr/solrj/src/java/org/apache/solr/common/util/CompressionUtil.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.common.util;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.zip.DataFormatException;
+import java.util.zip.Deflater;
+import java.util.zip.DeflaterInputStream;
+import java.util.zip.Inflater;
+
+/**
+ * Utility class for dealing with LZ4 compressed data, such as state.json in Zookeeper if configured to be compressed
+ */
+public class CompressionUtil {
+
+    /**
+     * Decompresses zlib compressed bytes, returning the uncompressed data as a byte[]
+     * @param data the input zlib compressed data to decompress
+     * @return the decompressed bytes
+     * @throws DataFormatException - The data is not zlib compressed data
+     */
+    public static byte[] decompressCompressedBytes(byte[] data) throws DataFormatException {
+        if (data == null) return null;
+        Inflater inflater = new Inflater();
+        inflater.setInput(data);
+        byte[] buf = new byte[100];
+        int read;
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
```

Comment:
the no-arg ctor creates a ByteArrayOutputStream backed by a default byte[] of length 32, resizes by doubling/copying. We could without much trouble estimate the size (we know the total size of the data compressed, and we know (roughly) the expected compression ratio, and we know that we only compress if the uncompressed size is > `MIN_BYTES_FOR_COMPRESSION`).

That said, I'd actually advocate for a different approach: we could write our own "header" including magic bytes that would allow us to identify the stream as compressed (it's possible that Zlib already does this) -- but more importantly we could use our header to record the exact size number of bytes of the decompressed payload. This would allow us to:
1. skip creating a DataFormatException, filling in stack trace, etc. only to throw it away
2. skip BAOS and its inevitable (even if proper initial size) arraycopy on final `toByteArray()`